### PR TITLE
add unit tests for converters package

### PR DESCRIPTION
**Pull request by sonasingh46**
Signed-off-by: Ashutosh Kumar <sonasingh46@gmail.com>

**What this PR does / why we need it**:
Adds unit tests for missing functions in converters package.
Increases the test coverage from 60% to 92.7%.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
None
```

### CODE REVIEW

File: azure/converters/vm_test.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+	. "github.com/onsi/gomega"
+	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
+)
+
+func Test_SDKToVM(t *testing.T) {
+	cases := []struct {
+		name      string
+		sdkVM     compute.VirtualMachine
+		expect    func(*GomegaWithT, *VM)
+		expectErr bool
```
Comment (Jont828) [inline azure/converters/vm_test.go:33]: Would this function ever return an error?

File: azure/converters/image_test.go [hunk 2]
```diff
@@ -26,6 +27,96 @@ import (
...
+			expect: func(g *GomegaWithT, result *compute.ImageReference) {
+				ID := to.StringPtr(fmt.Sprintf("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Compute/galleries/%s"+
+					"/images/%s/versions/%s", "fake-sub-id", "fake-rg", "fake-gallery-name", "fake-image-name", "v1.0.0"))
+				g.Expect(result.ID).To(Equal(ID))
+			},
+			expectErr: false,
+		},
+		{
+			name:      "Should return error if image is not specified",
+			expectErr: true,
```
Comment (Jont828) [inline azure/converters/image_test.go:95]: Wdyt about using a string for the expected error message instead of a bool? We do that [here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/azure/services/inboundnatrules/spec_test.go). The `expectedErr` string let's us check to verify the error message is correct, and we can set it to `""` to represent no error.

```
if tc.expectedError != "" {
	g.Expect(err).To(HaveOccurred())
	g.Expect(err).To(MatchError(tc.expectedError))
} else {
	g.Expect(err).NotTo(HaveOccurred())
	g.Expect(res).To(Equal(tc.expectedPortResult))
}
```

File: azure/converters/rules_test.go [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+			securityRule: infrav1.SecurityRule{
+				Name:             "fake-rule",
+				Description:      "fake rule description",
+				Source:           to.StringPtr("fake-source"),
+				SourcePorts:      to.StringPtr("fake-source-ports"),
+				Destination:      to.StringPtr("fake-destination"),
+				DestinationPorts: to.StringPtr("fake-destination-port-ranges"),
+				Priority:         2,
+				Protocol:         infrav1.SecurityGroupProtocolAll,
+				//Direction: infrav1.SecurityRuleDirectionOutbound,
```
Comment (Jont828) [inline azure/converters/rules_test.go:45]: Remove comment? I think it's fine to not test combinations of `rule.Protocol` and `rule.Direction` here since it'll have way too many combinations.

File: azure/converters/spotinstances.go [hunk 4]
```diff
@@ -36,6 +37,7 @@ func GetSpotVMOptions(spotVMOptions *infrav1.SpotVMOptions) (compute.VirtualMach
 		if err != nil {
 			return "", "", nil, err
 		}
+		fmt.Println(err)
```
Comment (Jont828) [inline azure/converters/spotinstances.go:40]: Remove print statement?

File: azure/converters/spotinstances_test.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+		billingProfile        *compute.BillingProfile
+	}
+	cases := []struct {
+		name          string
+		spotVMOptions *infrav1.SpotVMOptions
+		expectErr     bool
+		expect        func(*GomegaWithT, resultParams)
+	}{
+		{
+			name:          "should return vm priority, eviction policy and nil billing profile for spotVMOptions",
```
Comment (Jont828) [inline azure/converters/spotinstances_test.go:42]:
```suggestion
			name:          "should return vm priority, eviction policy and nil billing profile for spotVMOptions with nil MaxPrice",
```

File: azure/converters/spotinstances_test.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+			spotVMOptions: &infrav1.SpotVMOptions{},
+			expectErr:     false,
+			expect: func(g *GomegaWithT, result resultParams) {
+				g.Expect(result.vmPriorityTypes).To(Equal(compute.VirtualMachinePriorityTypesSpot))
+				g.Expect(result.vmEvictionPolicyTypes).To(Equal(compute.VirtualMachineEvictionPolicyTypesDeallocate))
+				g.Expect(result.billingProfile).Should(BeNil())
+			},
+		},
+		{
+			name: "should return vm priority, eviction policy and billing profile for spotVMOptions",
```
Comment (Jont828) [inline azure/converters/spotinstances_test.go:52]:
```suggestion
			name: "should return vm priority, eviction policy and billing profile for spotVMOptions with set MaxPrice",
```

File: azure/converters/spotinstances_test.go (hunk 5 above)
Comment (Jont828) [inline azure/converters/spotinstances_test.go:42]: Can we add a test for when `spotVMOptions == nil` for the base case?

File: azure/converters/image_test.go [hunk 7]
```diff
@@ -26,6 +27,96 @@ import (
...
+				ID: to.StringPtr("fake-id"),
+				SharedGallery: &infrav1.AzureSharedGalleryImage{
+					SubscriptionID: "fake-sub-id",
+					ResourceGroup:  "fake-rg",
+					Gallery:        "fake-gallery-name",
+					Name:           "fake-image-name",
+					Version:        "v1.0.0",
+				},
+			},
+			expect: func(g *GomegaWithT, result *compute.ImageReference) {
```
Comment (Jont828) [inline azure/converters/image_test.go:48]: Wdyt about changing the `expect` part to `expectedResult`? That way, you can directly compare the expected conversion result in w/ something like this `g.Expect(res).To(Equal(tc.expectedResult))` in `t.Run()` and don't need to write it in each test case. Check out [this](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/azure/services/inboundnatrules/spec_test.go) for an example.

File: azure/converters/tags_test.go [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+func TestMapToTags(t *testing.T) {
+	tests := []struct {
+		name   string
+		srcMap map[string]*string
+		want   v1beta1.Tags
+	}{
+		{
+			name: "convert map to tags",
+			srcMap: map[string]*string{
```
Comment (Jont828) [inline azure/converters/tags_test.go:35]: Although this is really simple, can we add a test for (1) the map is empty, and (2) what if the map value is a nil string pointer?

**Review by Jont828 — COMMENTED** [review]
I like that we're adding test cases here! Just a couple comments about how we could handle the results/errors.

File: azure/converters/vm_test.go (hunk 1 above)
Comment (sonasingh46) [reply azure/converters/vm_test.go:33]: This should not -- but that mean this functions needs refactoring.  Will do that.

File: azure/converters/image_test.go (hunk 7 above)
Comment (sonasingh46) [reply azure/converters/image_test.go:48]: Either way should be fine. In all the test case -- we have to any way write the expected `infrav1.AzureSharedGalleryImage` struct. The only difference here is that it wrapped by a function. In fact, I thought about it while writing but it has been written the same in the existing code so did not break the pattern. 
Anyways, if you feel to go with the suggested approach -- I can make that change.

**Comment by sonasingh46** [comment]
cc @CecileRobertMichon @jackfrancis @mboersma

File: azure/converters/image_test.go (hunk 7 above)
Comment (Jont828) [reply azure/converters/image_test.go:48]: I got the `expectedResult` convention from the other tests like in `cluster_test.go`(https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/azure/scope/cluster_test.go). IMO I prefer that method since it checks every field and is easier to understand at a glance. Wdyt @CecileRobertMichon

File: azure/converters/spotinstances_test.go [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+	cases := []struct {
+		name          string
+		spotVMOptions *infrav1.SpotVMOptions
+		expectErr     bool
+		expect        func(*GomegaWithT, resultParams)
+	}{
+		{
+			name:          "should return empty vm priority, empty eviction policy and nil billing profile for nil spotVMOptions",
+			spotVMOptions: nil,
+			expectErr:     false,
```
Comment (Jont828) [inline azure/converters/spotinstances_test.go:44]: Can we use `expectErr` as a string like in image_test.go and add a test case where we are returning an error?

File: azure/converters/tags_test.go [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+
+func TestMapToTags(t *testing.T) {
+	tests := []struct {
+		name   string
+		srcMap map[string]*string
+		want   v1beta1.Tags
+	}{
+		{
+			name:   "convert nil map to tags",
+			srcMap: map[string]*string{},
```
Comment (Jont828) [inline azure/converters/tags_test.go:35]: Is there a difference between `nil` map and empty map, i.e. set `srcMap` to `nil`? If not then we could merge this and the empty map test into one.

File: azure/converters/image_test.go (hunk 7 above)
Comment (CecileRobertMichon) [reply azure/converters/image_test.go:48]: +1 to using `expectedResult` if we're trying to compare the entire struct (in this case looks like we are). In some other tests it's sometimes useful to have an `expect` func so we can customize the checks a bit more by test case and only check a specific desired outcome

**Comment by CecileRobertMichon** [comment]
/close

in favor of #2564 etc.

/cc @willie-yao
